lib/internal/modules/esm/assets/load.js

```diff
@@ -0,0 +1,241 @@
+'use strict';
+const {
+  globalThis,
+  ArrayFrom,
+  ArrayPrototypeJoin,
+  ArrayPrototypePush,
+  ArrayPrototypeSplice,
+  Promise,
+  PromiseResolve,
+  SafeMap,
+  StringPrototypeSlice,
+  StringPrototypePadStart,
+  StringPrototypeSplit,
+  StringPrototypeStartsWith,
+} = primordials;
+const {
+  Buffer: {
+    concat: BufferConcat
+  }
+} = globalThis;
+const {
+  codes: {
+    ERR_UNSUPPORTED_ESM_NETWORK_IMPORT_SPECIFIER,
+    ERR_UNSUPPORTED_ESM_NETWORK_BAD_RESPONSE
+  }
+} = require('internal/errors');
+const { URL } = require('internal/url');
+
+
+/**
+ * Lazily load these to avoid snapshot issues
+ */
+
+function HTTPSGet(url, opts) {
+  // eslint-disable-next-line no-func-assign
+  HTTPSGet = require('https').get;
+  return HTTPSGet(url, opts);
+}
+
+function HTTPGet(url, opts) {
+  // eslint-disable-next-line no-func-assign
+  HTTPGet = require('http').get;
+  return HTTPGet(url, opts);
+}
+
+function dnsLookup(name, opts) {
+  // eslint-disable-next-line no-func-assign
+  dnsLookup = require('dns/promises').lookup;
+  return dnsLookup(name, opts);
+}
+
+/**
+ * @param {URL} parsed
+ * @returns {Promise<CacheEntry> | CacheEntry}
+ */
+function fetchWithRedirects(parsed) {
+  const existing = getCache.get(parsed.href);
+  if (existing) {
+    return existing;
+  }
+  const handler = parsed.protocol === 'http:' ? HTTPGet : HTTPSGet;
+  const result = new Promise((fulfill, reject) => {
+    const req = handler(parsed, {
+      headers: {
+        Accept: '*/*'
+      }
+    })
+    .on('error', reject)
+    .on('response', (res) => {
+      function dispose() {
+        req.destroy();
+        res.destroy();
+      }
+      if (res.statusCode >= 300 && res.statusCode <= 303) {
+        if (res.headers.location) {
+          dispose();
+          const location = new URL(res.headers.location, parsed);
+          if (location.protocol !== 'http:' && location.protocol !== 'https:') {
+            reject(new ERR_UNSUPPORTED_ESM_NETWORK_IMPORT_SPECIFIER(
+              res.headers.location,
+              parsed.href,
+              'cannot redirect to non-network location'));
+            return;
+          }
+          return PromiseResolve(
+            fetchWithRedirects(location)
```

Comment:
should we mitigate infinite redirect loops?

Reply:
yes, eventually